concordium-consensus/src/Concordium/ImportExport.hs

```diff
@@ -127,160 +191,123 @@ writeSection ::
     Handle ->
-    -- |Action to write out the blocks and return the number of them
-    m Word64 ->
-    -- |Action to write out the finalization records and return the number of them
-    m Word64 ->
+    -- |Chunk size
+    Word64 ->
     m ()
-writeSection
+writeChunks
     sectionGenesisIndex
     sectionProtocolVersion
     sectionGenesisHash
     sectionFirstBlockHeight
-    hdl
-    writeBlocks
-    writeFinRecs = do
+    outDir indexHdl chunkSize = do
+        let chunkName = outDir </> "blocks-"
+                        ++ show sectionGenesisIndex ++ "-"
+                        ++ show sectionFirstBlockHeight
+                        ++ ".dat"
+        chunkHdl <- liftIO $ initialHandle chunkName
+
         (sectionStart, blocksStart) <- liftIO $ do
-            sectionStart <- hTell hdl
+            BS.hPut chunkHdl $ encode (3 :: Version)
+            sectionStart <- hTell chunkHdl
             -- Write a dummy section header that we will later overwrite
-            runPutH (liftPut $ putWord64be sectionHeaderLength >> put placeholderSectionHeader) hdl
-            blocksStart <- hTell hdl
+            runPutH (liftPut $ putWord64be sectionHeaderLength >> put placeholderSectionHeader) chunkHdl
+            blocksStart <- hTell chunkHdl
             return (sectionStart, blocksStart)
-        sectionBlockCount <- writeBlocks
-        blocksEnd <- liftIO $ hTell hdl
-        sectionFinalizationCount <- writeFinRecs
+        -- how many blocks would it take to fill the exported chunk to its `chunkSize`
+        -- (non-zero means we're out of blocks to export in the current section)
+        (chunkUnused, finRecsMissing) <- exportBlocksToChunk chunkHdl sectionFirstBlockHeight chunkSize
+        let sectionBlockCount = chunkSize - chunkUnused
+        blocksEnd <- liftIO $ hTell chunkHdl
+        sectionFinalizationCount <- exportFinRecsForChunk chunkHdl finRecsMissing
         liftIO $ do
-            sectionEnd <- hTell hdl
+            sectionEnd <- hTell chunkHdl
             -- Go back to the start and rewrite the section header with the correct data
-            hSeek hdl AbsoluteSeek sectionStart
+            hSeek chunkHdl AbsoluteSeek sectionStart
             let sectionHeader =
                     SectionHeader
                         { sectionLength = fromInteger (sectionEnd - sectionStart),
                           sectionBlocksLength = fromInteger (blocksEnd - blocksStart),
                           ..
                         }
-            runPutH (liftPut $ putWord64be sectionHeaderLength >> put sectionHeader) hdl
-            hSeek hdl AbsoluteSeek sectionEnd
+            runPutH (liftPut $ putWord64be sectionHeaderLength >> put sectionHeader) chunkHdl
+            hClose chunkHdl
+            TIO.hPutStrLn indexHdl . T.pack $
+              takeFileName chunkName
+              ++ "," ++ show sectionFirstBlockHeight
+              ++ "," ++ show (sectionFirstBlockHeight + BlockHeight sectionBlockCount - 1)
```

Comment:
I'm not sure I understand. The `Show` instance for `BlockHeight` is used here because the argument of `show` has type `BlockHeight`.

Also, in https://github.com/Concordium/concordium-node/pull/440/#discussion_r938796601 you advise against using the `Show` instance of `BlockHeight`, contrarily to the current comment. If that advice holds, what would you suggest here to turn `BlockHeight`s into strings that can be `Read`?

Reply:
Sorry, this is a typo. I meant we should not rely on the `Show` instance for `BlockHeight`. i.e. convert them to a Word64 or similar and `show` that.